### PR TITLE
Add Output Variables to Ideal Air loads

### LINKED ISSUE #8284: Add New Zone Ideal Loads Output Variables
Issue overview
--------------
#8109 partially fixes #5788, but this comment should still be considered: https://github.com/NREL/EnergyPlus/issues/5788#issuecomment-235060715

TL;DR: add the following output vars:
- [x] Zone Ideal Loads Supply Air Temperature
- [x] Zone Ideal Loads Supply Air Humidity Ratio

### Details
Some additional details for this issue (if relevant):
 - Platform (All)
 - Version of EnergyPlus (17f452d0404464f6bfba11cfe44d67abf32fa379)
 - Unmethours link or helpdesk ticket number - None

### Checklist
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [ ] Defect file added (list location of defect file here)
 - [ ] Ticket added to Pivotal for defect (development team task)
 - [x] Pull request created (the pull request will have additional tasks related to reviewing changes that fix this defect)

### PR DESCRIPTION
**Pull request by jmythms**
Pull request overview
---------------------
 - Fixes #8284 Add New Zone Ideal Loads Output Variables
 - This was a request to add four new output variables to Ideal Air loads. The four new variables added are:
1. Zone Ideal Loads Supply Air Temperature
2. Zone Ideal Loads Supply Air Humidity Ratio
3. Zone Ideal Loads Mixed Air Temperature
4. Zone Ideal Loads Mixed Air Humidity Ratio

The documentation has also been changed. 

### Pull Request Author
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [x] Title of PR should be user-synopsis style (clearly understandable in a standalone changelog context)
 - [x] Label the PR with at least one of: Defect, Refactoring, NewFeature, Performance, and/or DoNoPublish
 - [x] Pull requests that impact EnergyPlus code must also include unit tests to cover enhancement or defect repair
 - [x] Author should provide a "walkthrough" of relevant code changes using a GitHub code review comment process
 - [x] If any diffs are expected, author must demonstrate they are justified using plots and descriptions
 - [ ] If changes fix a defect, the fix should be demonstrated in plots and descriptions
 - [ ] If any defect files are updated to a more recent version, upload new versions here or on DevSupport
 - [ ] If IDD requires transition, transition source, rules, ExpandObjects, and IDFs must be updated, and add IDDChange label
 - [ ] If structural output changes, add to output rules file and add OutputChange label
 - [ ] If adding/removing any LaTeX docs or figures, update that document's CMakeLists file dependencies

### Reviewer
This will not be exhaustively relevant to every PR.
 - [x] Perform a Code Review on GitHub
 - [ ] If branch is behind develop, merge develop and build locally to check for side effects of the merge
 - [ ] If defect, verify by running develop branch and reproducing defect, then running PR and reproducing fix
 - [ ] If feature, test running new feature, try creative ways to break it
 - [x] CI status: all green or justified
 - [x] Check that performance is not impacted (CI Linux results include performance check)
 - [x] Run Unit Test(s) locally
 - [ ] Check any new function arguments for performance impacts
 - [ ] Verify IDF naming conventions and styles, memos and notes and defaults
 - [x] If new/modified idf included, locally check the err file and other outputs

### CODE REVIEW

File: doc/input-output-reference/src/overview/group-zone-forced-air-units.tex
Comment (mitchute) [inline doc/input-output-reference/src/overview/group-zone-forced-air-units.tex]: We should use the ```\SI``` package for units.

File: doc/input-output-reference/src/overview/group-zone-forced-air-units.tex
Comment (mitchute) [inline doc/input-output-reference/src/overview/group-zone-forced-air-units.tex:504]: These units are fine, though 🙈

File: src/EnergyPlus/PurchasedAirManager.cc
Comment (mitchute) [inline src/EnergyPlus/PurchasedAirManager.cc]: This is just passing the inlet temp and hum-rat through. I believe we want to use the already existing local variables ```SupplyTemp``` and ```SupplyHumRat``` to set these values.

File: src/EnergyPlus/PurchasedAirManager.hh
Comment (mitchute) [inline src/EnergyPlus/PurchasedAirManager.hh]: I'd shorten these to just ```SupplyAirTemp``` and ```SupplyHumRat```.

**Review by mitchute — CHANGES_REQUESTED** [review]
This will need a few minor cleanups. But it should be ready after that.

File: src/EnergyPlus/PurchasedAirManager.cc
Comment (mjwitte) [reply src/EnergyPlus/PurchasedAirManager.cc]: Or would it work to just move `SupplyTemp` and `SupplyHumRat` from local into the `PurchAir` struct?

Comment (mjwitte) [reply src/EnergyPlus/PurchasedAirManager.cc]: And while you're in here, `MixedAirTemp` and `MixedAirHumRat` would also be useful outputs.

Comment (mitchute) [reply src/EnergyPlus/PurchasedAirManager.cc]: That's fine with me.

File: doc/input-output-reference/src/overview/group-zone-forced-air-units.tex
Comment (jmythms) [reply doc/input-output-reference/src/overview/group-zone-forced-air-units.tex]: Shall I change all of the title lines (like line 493) to use the \SI package? Or just in the text portion, like line 495?

File: src/EnergyPlus/PurchasedAirManager.cc
Comment (jmythms) [reply src/EnergyPlus/PurchasedAirManager.cc]: Will do!

File: doc/input-output-reference/src/overview/group-zone-forced-air-units.tex
Comment (mitchute) [reply doc/input-output-reference/src/overview/group-zone-forced-air-units.tex]: Just change the in-text units. In this case, just move to ```\si{\celsius}``` (or whatever it ends up being). If you see others in this file, feel free to change them.

Comment (jmythms) [reply doc/input-output-reference/src/overview/group-zone-forced-air-units.tex]: Got it, thanks!

File: doc/input-output-reference/src/overview/group-zone-forced-air-units.tex
Comment (jmythms) [inline doc/input-output-reference/src/overview/group-zone-forced-air-units.tex:514]: New addition. Does it look ok?

File: doc/input-output-reference/src/overview/group-zone-forced-air-units.tex
Comment (jmythms) [inline doc/input-output-reference/src/overview/group-zone-forced-air-units.tex:59]: Mostly changing the tex code to use the siunitx package.

File: testfiles/HVACTemplate-5ZonePurchAir.idf
Comment (jmythms) [inline testfiles/HVACTemplate-5ZonePurchAir.idf:1942]: Now the results appear in the test file output. Should I change the reporting frequency to hourly? Also, the effects are better visible for an annual simulation.

File: src/EnergyPlus/PurchasedAirManager.cc
Comment (jmythms) [inline src/EnergyPlus/PurchasedAirManager.cc]: Not sure if they need to be reset, and if 0.0 is ok.

File: src/EnergyPlus/PurchasedAirManager.hh
Comment (jmythms) [inline src/EnergyPlus/PurchasedAirManager.hh:200]: Removed them since I felt the new mixedairtemp and mixedairhumrat were doing the same thing?

**Comment by mitchute** [comment]
@jmythms I'm not sure what the deal is with the regressions. Maybe we hit an API limit? (FYI @Myoldmopar ) Regardless of that, there are still some build warnings that will need to be addressed. Let me know if you need help figuring those out.

**Comment by nrel-bot-2** [comment]
@jmythms @lgentile it has been 28 days since this pull request was last updated.

**Comment by jmythms** [comment]
More details:
image
image

Where is the heat recovered from? Is it the recirc/return air ([equation](https://github.com/NREL/EnergyPlus/blob/283b64077bc031c2b6be032380370b8f2c8e6cac/src/EnergyPlus/PurchasedAirManager.cc#L2956))? If so, will the return/recirc air properties change as well?

**Comment by mjwitte** [comment]
@jmythms There is a relief air stream not shown in these diagrams which would be air from B that is exhausted to balance the incoming outdoor air (C). So, the heat recovery is between B and C and the flow rate of C.

**Comment by mjwitte** [comment]
To answer the second question - yes and no. The relief air properties change, but that air is exhausted so it's not relevant. The portion of B that mixes with the OA from C does not change properties.

**Comment by jmythms** [comment]
@mjwitte Thanks a lot!

> To answer the second question - yes and no. The relief air properties change, but that air is exhausted so it's not relevant. The portion of B that mixes with the OA from C does not change properties.

Thanks a lot!!

**Comment by nrel-bot-3** [comment]
@jmythms @lgentile it has been 28 days since this pull request was last updated.

File: src/EnergyPlus/PurchasedAirManager.hh
Comment (mjwitte) [reply src/EnergyPlus/PurchasedAirManager.hh:200]: And it appears that  `FinalMixedAirTemp` and `FinalMixedAirHumRat` weren't being used anywhere, so find to replace them here.

File: testfiles/HVACTemplate-5ZonePurchAir.idf
Comment (mjwitte) [reply testfiles/HVACTemplate-5ZonePurchAir.idf:1942]: Yes, I would make the reporting frequency consistent - looks like everything else is hourly.

**Review by mjwitte — COMMENTED** [review]
@jmythms Sorry this fell in the abyss over the holidays.
This needs a unit test and there's a conflict now with develop - should be an easy merge/fix.
A couple of other preliminary review comments below.

**Comment by jmythms** [comment]
> @jmythms Sorry this fell in the abyss over the holidays.
> This needs a unit test and there's a conflict now with develop - should be an easy merge/fix.
> A couple of other preliminary review comments below.

Thanks for the review! Getting to work on the changes now.

**Comment by jmythms** [comment]
@mjwitte May I get a review on this PR please? I added a new unit test for the new output variables.

File: doc/input-output-reference/src/overview/group-zone-forced-air-units.tex
Comment (rraustad) [reply doc/input-output-reference/src/overview/group-zone-forced-air-units.tex:514]: Yes, this looks reasonable. I thought about adding "entering the zone inlet node" to be more specific but with the change for zone/spaces coming it's probably better at this point to leave this generic.

File: src/EnergyPlus/PurchasedAirManager.hh
Comment (rraustad) [inline src/EnergyPlus/PurchasedAirManager.hh]: Not sure why this line didn't wrap earlier. It's much longer than all the rest. Was Clang format used?

**Comment by rraustad** [comment]
@mjwitte I have reviewed and this looks good. I would expect 0 diff's with this change so as long as CI shows that it should be ready to merge (save any last minute edits). Also, these new report variables, do those need to be included in a new report variable summary for users?

**Comment by mjwitte** [comment]
@rraustad We don't normally list new outputs in the output changes doc, maybe we should, but I wouldn't hold up this one. Did you reiew I/O Ref pdf to make sure all the changes are ok? Feel free to complete this review and merge.

**Comment by rraustad** [comment]
OK, I'll build the docs and run the example file. I see MTD and big table diff's on HVACTemplate-5ZonePurchAir. @jmythms did you review those diff's?

File: src/EnergyPlus/PurchasedAirManager.hh
Comment (jmythms) [reply src/EnergyPlus/PurchasedAirManager.hh]: @rraustad I just checked this, it was disabled in my IDE. Shall I do one more commit after reformatting?

Comment (rraustad) [reply src/EnergyPlus/PurchasedAirManager.hh]: @Myoldmopar is it worth another commit for Clang formatting the hh file? I don't see any other changes needed, but I haven't tested the example file yet. Unit tests run clean locally.

**Comment by jmythms** [comment]
> OK, I'll build the docs and run the example file. I see MTD and big table diff's on HVACTemplate-5ZonePurchAir. @jmythms did you review those diff's?

The diffs were generated because I asked for the new variables in the IDF file. 

The MTD diffs are changes in meter numbers, screenshot attached below. 
image
image

Big table diffs are changes to the input fields, which were created from asking for these new variables:
image
image

File: doc/input-output-reference/src/overview/group-zone-forced-air-units.tex
Comment (rraustad) [inline doc/input-output-reference/src/overview/group-zone-forced-air-units.tex:4602]: There is an error here in the units (i.e., 0.012 C). The existing units were converted correctly, however, the units should be kgWater/kgDryAir. Also, the previous sentence uses kg/kg, these units could also be updated to be consistent. This is the only issue I found with the doc changes.

Comment (rraustad) [reply doc/input-output-reference/src/overview/group-zone-forced-air-units.tex:4602]: @jmythms there is a correction needed in the doc. This gives you the opportunity to Clang format the hh as needed (and cc? but I didn't see anything there). Give me a little time for more review before you make a commit.

Comment (jmythms) [reply doc/input-output-reference/src/overview/group-zone-forced-air-units.tex:4602]: That was a stupid mistake on my part. I will make the changes, and add the reformatted "src/EnergyPlus/PurchasedAirManager.hh" file in with that commit.

**Comment by rraustad** [comment]
Results using new example file. This looks good. SA humidity ratio is unchanged in heating mode and always lower than mixed air in cooling mode.

image

**Comment by rraustad** [comment]
@jmythms I have finished my review. Go ahead and make your changes and commit.

**Review by rraustad — CHANGES_REQUESTED** [review]
Reviewed code, docs and unit test. With the next commit this is ready for merge.

**Comment by rraustad** [comment]
@mjwitte @Myoldmopar this is ready to go after this next commit. I will merge when I can, just acknowledging my review and that this is approved and good to go.

**Comment by jmythms** [comment]
> @jmythms I have finished my review. Go ahead and make your changes and commit.

Thank you so much for the review!

**Comment by rraustad** [comment]
@jmythms also check off the boxes in the Pull request overview above under Pull request author. I could do that but I would like you to start doing that for your PRs. It's a reminder check against our work so we don't forget anything.

File: doc/input-output-reference/src/overview/group-zone-forced-air-units.tex
Comment (jmythms) [reply doc/input-output-reference/src/overview/group-zone-forced-air-units.tex:4602]: Fixed this:
image

**Comment by jmythms** [comment]
> @jmythms also check off the boxes in the Pull request overview above under Pull request author. I could do that but I would like you to start doing that for your PRs. It's a reminder check against our work so we don't forget anything.

I have checked it for this PR and will start doing it for the other ones. The requested changes were made. Thanks again.

I also  accidentally clang-formatted the "ScheduleManager.hh" file in the [863baf5](https://github.com/NREL/EnergyPlus/pull/8302/commits/863baf54c9f5fc84c3161eb3d5ed15b57174bf2b) commit, did not revert it since I thought leaving it formatted would be a good thing. Just wanted to let you know.

**Comment by rraustad** [comment]
@jmythms, a tip moving forward. I want you to notice how much has changed when you clang format a file. I am glad I did not have to review the code with these format changes included with yours. The reviewer would have a very difficult time finding the actual code changes. Either format only the code you touch (in MSVS highlight the code, select Edit on the menu, then Advanced, then format selection) or clang format the file after reviewers have a chance to look at your changes. Just a tip to make the reviewers life easier. Anyway, I think this is ready now and I will run the unit tests one more time and then this can be merged.

**Comment by rraustad** [comment]
Unit tests run clean locally as expected.

**Comment by rraustad** [comment]
There is another diff here, the SolarShadingTestGPU example file but I think that has caused this same diff for some time.

**Comment by rraustad** [comment]
@jmythms did you have a pivotal ticket for this? I can't find one.

**Pull request merged**
